test/src/unit-cppapi-bitsort-filter.cc

```diff
@@ -0,0 +1,580 @@
+  read_query.add_range("x", dim_lo, dim_hi);
+
+  if (num_dims >= 2) {
+    read_query.add_range("y", dim_lo, dim_hi);
+  }
+
+  if (num_dims == 3) {
+     read_query.add_range("z", dim_lo, dim_hi);
+  }
+}
+
+/**
+ * @brief Tests the bitsort filter, given the parameters passed in.
+ * This function creates an array with num_dims dimensions and 
+ * the bitsort filter. Then, it writes into this array with randomly
+ * generated attribute data, and then reads from the array to test
+ * the bitsort filter.
+ *
+ * @tparam AttrType The type of the attribute data in the array.
+ * @tparam DimType The type of the dimensions of the array.
+ * @tparam AttributeDistribution The type of the random distribution used to
+ * generate the attribute data.
+ * @param bitsort_array_name The array name.
+ * @param num_dims The number of dimensions.
+ * @param write_layout The layout of the write query.
+ * @param read_layout The layout of the read query.
+ * @param set_subarray Whether the read query is called with a subarray encompassing the whole array.
+ */
+template <typename AttrType, typename DimType, typename AttributeDistribution>
+void bitsort_filter_api_test(
+    const std::string& bitsort_array_name,
+    uint64_t num_dims,
+    tiledb_layout_t write_layout,
+    tiledb_layout_t read_layout,
+    bool set_subarray) {
+  // Setup.
+  Context ctx;
+  VFS vfs(ctx);
+
+  if (vfs.is_dir(bitsort_array_name)) {
+    vfs.remove_dir(bitsort_array_name);
+  }
+
+  Domain domain(ctx);
+
+  // Add first dimension.
+  REQUIRE(num_dims >= 1);
+  REQUIRE(num_dims <= 3);
+  size_t num_element_per_dim = (BITSORT_DIM_HI - BITSORT_DIM_LO + 1);
+  size_t number_elements = num_element_per_dim;
+  domain.add_dimension(Dimension::create<DimType>(
+      ctx,
+      "x",
+      {{static_cast<DimType>(BITSORT_DIM_LO), static_cast<DimType>(BITSORT_DIM_HI)}},
+      static_cast<DimType>(TILE_EXTENT)));
+
+  if (num_dims >= 2) {
+    domain.add_dimension(Dimension::create<DimType>(
+        ctx,
+        "y",
+        {{static_cast<DimType>(BITSORT_DIM_LO), static_cast<DimType>(BITSORT_DIM_HI)}},
+        static_cast<DimType>(TILE_EXTENT)));
+    number_elements *= num_element_per_dim;
+  }
+  if (num_dims == 3) {
+    domain.add_dimension(Dimension::create<DimType>(
+        ctx,
+        "z",
+        {{static_cast<DimType>(BITSORT_DIM_LO), static_cast<DimType>(BITSORT_DIM_HI)}},
+        static_cast<DimType>(TILE_EXTENT)));
+    number_elements *= num_element_per_dim;
+  }
+
+  Filter f(ctx, TILEDB_FILTER_BITSORT);
+  FilterList filters(ctx);
+  filters.add_filter(f);
+
+  auto a = Attribute::create<AttrType>(ctx, "a");
+  a.set_filter_list(filters);
+
+  ArraySchema schema(ctx, TILEDB_SPARSE);
+  schema.set_domain(domain);
+  schema.add_attribute(a);
+  Array::create(bitsort_array_name, schema);
+
+  // Setting up the random number generator for the bitsort filter testing.
+  std::mt19937_64 gen(SEED);
+  AttrType low = 0;
+  if constexpr (std::is_signed<AttrType>::value) {
```

Comment:
Is this if statement necessary? I think ::min() should return 0 for unsigned types.

Reply:
Also I'm not sure if AttrType can be a floating point type? If so, you might consider using ::lowest()